File: src/main/java/com/intetics/lukyanenko/dao/CommonDAO.java
package com.intetics.lukyanenko.dao;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

public interface CommonDAO<T>
{
  List<T> findAll();
  void add(T model);
  void update(T model);
  void delete(T model);
  int generateNewID();
  
  T getByID(int id);
  T populateFromResultSet(ResultSet resultSet, String fieldNamePrefix) throws SQLException;
}

File: src/main/java/com/intetics/lukyanenko/dao/CommonDAO.kt
package com.intetics.lukyanenko.dao

import java.sql.ResultSet
import java.sql.SQLException

interface CommonDAO<T> {
    fun findAll(): List<T>
    fun add(model: T)
    fun update(model: T)
    fun delete(model: T)
    fun generateNewID(): Int

    fun getByID(id: Int): T?

    @Throws(SQLException::class)
    fun populateFromResultSet(resultSet: ResultSet, fieldNamePrefix: String): T
}
